specs/lightclient/beacon-chain.md

```diff
@@ -516,23 +521,130 @@ def process_justification_and_finalization(state: BeaconState) -> None:
+    process_rewards(state)
+    process_penalties(state)
+
+def process_rewards(state: BeaconState) -> None:
     # No rewards are applied at the end of `GENESIS_EPOCH` because rewards are for work done in the previous epoch
     if get_current_epoch(state) == GENESIS_EPOCH:
         return
-    flag_deltas = [get_flag_deltas(state, flag, numerator) for (flag, numerator) in get_flags_and_numerators()]
-    deltas = flag_deltas + [get_inactivity_penalty_deltas(state)]
-    for (rewards, penalties) in deltas:
+    flag_rewards = [get_flag_rewards(state, flag, numerator) for (flag, numerator) in get_flags_and_numerators()]
+    for rewards in flag_rewards:
         for index in range(len(state.validators)):
             increase_balance(state, ValidatorIndex(index), rewards[index])
-            decrease_balance(state, ValidatorIndex(index), penalties[index])
 ```
 
+```python
+def process_penalties(state: BeaconState) -> None:
+    # These validators participated in FFG and are exempt from the leak
+    matching_target_attesting_indices = get_unslashed_participating_indices(
+        state, TIMELY_TARGET_FLAG, get_previous_epoch(state)
+    )
+    
+    # A validator's leak_score updates as follows:
+    # (i)  If there was an inactivity leak in a given epoch, anyone who participated in that epoch
+    #      has their score decrease by 1, anyone who did not has their score increase by LEAK_SCORE_BIAS
+    # (ii) If there was not an inactivity leak in a given epoch, anyone who participated in that epoch
+    #      has their score decrease by 1
+    # The code below implements this, though amortizing the "increase if absent" component until the end
+    # of the ACTIVATION_EXIT_PERIOD
+    
+    if is_in_inactivity_leak(state):
+        leak_score_decrease = LEAK_SCORE_BIAS + 1
+        state.leak_epoch_counter += 1
+    else:
+        leak_score_decrease = 1
+    for index in get_eligible_validator_indices(state):
+        if index in matching_target_attesting_indices:
+            if state.leak_score[index] < leak_score_decrease:
+                state.leak_score[index] = 0
+            else:
+                state.leak_score[index] -= leak_score_decrease
+    # Penalty processing
+    if get_current_epoch(state) % EPOCHS_PER_ACTIVATION_EXIT_PERIOD == 0:
+        attestation_numerators = (TIMELY_HEAD_NUMERATOR, TIMELY_SOURCE_NUMERATOR, TIMELY_TARGET_NUMERATOR)
+        for index in get_eligible_validator_indices(state):
+            # "Regular" penalty
+            accumulated_penalty = get_base_reward(state, index) * sum(attestation_numerators) // REWARD_DENOMINATOR
+            decrease_balance(state, ValidatorIndex(index), accumulated_penalty)
+            # Inactivity-leak-specific penalty
+            if state.leak_score[index] >= EPOCHS_PER_ACTIVATION_EXIT_PERIOD * LEAK_SCORE_BIAS
+                leak_penalty = state.leak_score[index] * EPOCHS_PER_ACTIVATION_EXIT_PERIOD // LEAK_SCORE_BIAS // INACTIVITY_PENALTY_QUOTIENT
```

Comment:
Should this be multiplied by `(EPOCHS_PER_ACTIVATION_PERIOD - leak_epoch_counter)` to account for a partial leak during the period?